src/realm/group.hpp

```diff
@@ -968,49 +969,43 @@ inline ConstTableRef Group::get_table(StringData name) const
     return ConstTableRef(table, table ? table->m_alloc.get_instance_version() : 0);
 }
 
-inline TableRef Group::add_table(StringData name)
+inline TableRef Group::add_table(StringData name, Table::Type table_type)
 {
     if (!is_attached())
         throw LogicError(LogicError::detached_accessor);
     check_table_name_uniqueness(name);
-    Table* table = do_add_table(name, false); // Throws
+    Table* table = do_add_table(name, table_type); // Throws
     return TableRef(table, table->m_alloc.get_instance_version());
 }
 
-inline TableRef Group::add_embedded_table(StringData name)
-{
-    if (!is_attached())
-        throw LogicError(LogicError::detached_accessor);
-    check_table_name_uniqueness(name);
-    Table* table = do_add_table(name, true); // Throws
-    return TableRef(table, table->m_alloc.get_instance_version());
-}
-
-inline TableRef Group::get_or_add_table(StringData name, bool* was_added)
+inline TableRef Group::get_or_add_table(StringData name, Table::Type table_type, bool* was_added)
 {
+    REALM_ASSERT(table_type != Table::Type::Embedded);
     if (!is_attached())
         throw LogicError(LogicError::detached_accessor);
     auto table = do_get_table(name);
     if (was_added)
         *was_added = !table;
     if (!table) {
-        table = do_add_table(name, false);
+        table = do_add_table(name, table_type);
     }
     return TableRef(table, table->m_alloc.get_instance_version());
```

Comment:
Let's suppose we are going to rely on the sync protocol version to prevent creation of asymmetric tables. Now we set up the following scenario:
1. Create a Realm with an asymmetric table with the newest version of Realm on sync protocol 6. This works as expected.
2. Open the Realm again with an older version of the SDK that is running sync protocol 5 or 4. Make some writes to the asymmetric table while offline. The table is treated as a normal TopLevel table and the objects persist locally.
3. Open the Realm again with the latest version of the SDK on sync protocol 6. Connection is restored, and the changesets produced in the previous step while offline are picked up and synced. The server should be fine with this, but the client will be left with the objects persisted locally as well which is inconsistent. 

I think this can be avoided, by adding a check here in `get_or_add_table` where the table is cleared if it is asymmetric. That should wipe out the unwanted state made by older versions without requiring the file format bump.